### DISCUSSION TITLE
Ensemble model

### DISCUSSION THREAD

**Discussion by mw-boop** [ultralytics/ultralytics#21301] [Q&A] [answered]
Hi everyone,
I've got 10 yolo11 small segmentation models trained on the same dataset but with different hyperparameters. I've tried, and failed, to make an ensemble model using weight averaging. Is it possible to make an ensemble model with yolo11? 
kind regards

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mw-boop, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance—especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, which cover many common use cases and questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug your issue.

If your question is about custom training or model ensembling ❓, please share additional details such as your approach, relevant code snippets, and any errors or results you’ve observed. This will help us better understand your use case and provide more targeted assistance.

You can also connect with the Ultralytics community for support and discussion:
- [Discord](https://discord.com/invite/ultralytics) for real-time chat 🎧
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://www.reddit.com/r/Ultralytics/) for community Q&A

## Upgrade

Please ensure you’re using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response—an Ultralytics engineer will review your discussion and assist you soon!

**Reply by glenn-jocher** [reply, marked as answer]
Yes, you can create an ensemble model with YOLO11! Ultralytics provides an [`Ensemble` class](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.Ensemble) that allows combining multiple models. Here's how to use it with your 10 segmentation models:

```python
from ultralytics.nn.tasks import Ensemble
from ultralytics import YOLO

# Load your 10 trained models
model_paths = ['model1.pt', 'model2.pt', 'model3.pt', ...]  # your 10 model paths
models = [YOLO(path) for path in model_paths]

# Create ensemble
ensemble = Ensemble()
for model in models:
    ensemble.append(model.model)

# Use the ensemble for inference
results = ensemble(image_tensor)
```

The ensemble concatenates predictions from all models and can provide better performance than individual models, especially when the models have different strengths from varied hyperparameters.
